### ISSUE TITLE
Contracts breaks profiling

### ISSUE THREAD

**Issue by egonSchiele** [egonSchiele/contracts.ruby#182]
From this comment by Jeff Dickey: http://julienblanchard.com/2015/contracts-with-ruby/#comment-2136176415

> Contracts.ruby performs, in effect, C++-style name mangling on your methods; there's even a section of the tutorial that starts out, "You can use contracts for method overloading!" The way this works, of course, is that your multiple-methods-with-the-same-name don't really have the same name anymore. This breaks what turns out to be an important part of Ruby. We didn't figure this out until we tried adding the sandi_meter to our code, and it would fail because it couldn't examine our code — all it saw were the mangling wrappers that Contracts.ruby wrapped around our code.

**Comment by waterlink** [comment]
Hmm, we do include original method name in generated method:

```ruby
__contracts_ruby_#{original_method_name}_#{random_id_for_uniqueness}
```

I assume it is not readable enough to spot which method you need to optimize :(

Can we try to make method name first and only after that all the other stuff? Like this:

```ruby
#{original_method_name}__contracts_ruby_#{random_id_for_uniqueness}
```

WDYT?

/cc @jdickey

**Comment by waterlink** [comment]
Other idea, is that when using pattern matching, you might want to know which overload exactly causes the slowdown. For that we need to somehow include this information in mangling method name.. Idea:

```ruby
Contract Num, String => Num
def something; end

Contract Num, Num, String => Num
def something; end
```

will render such method names:

```ruby
"something__Contract_Num_String_Num__4df6sd4asf56asdf4"
"something__Contract_Num_Num_String_Num__238471235jh15hjg"
```

Here I have serialized contract value to valid identifier (while dropping all stuff that can't be part of identifier) and put it there.

Might work or might not. Might be worth it or might not. WDYT?

**Comment by waterlink** [comment]
@jdickey can you provide some more insight on the issue? Or even to provide some example to play around and **feel** the problem.

**Comment by waterlink** [comment]
Heh, I just played around, and it is quite possible to do these names:

```ruby
"something Contract=[Num, String => Num] id=67s5f6sd6f56sda75"
"something Contract=[Num, Num, String => Num] id=223hjf41212h3213gf"
```

They are perfect valid methods (at least in ruby 2.2, didn't test in other rubies). Of course you can't call them directly, but you can `#send` them, which is fine, since they are getting called only through `#send` and they are implementation detail of contracts.

This should make them extremely readable in profiler output. WDYT?

**Comment by waterlink** [comment]
Ah, I see. This will make it more readable for performance profiler output, but will not really help with static analyzers :(

It is probably possible to run static analyzer with `NO_CONTRACTS=true`. Then you still have a problem with pattern-matching contracts, since they ignore this option.

**Comment by egonSchiele** [comment]
Ok, responded to @jdickey on the thread too, curious to know what he thinks. Those names seem good to me.

**Comment by jdickey** [comment]
Guys, I think it's only sensible that the original method be accessible (if aliased as @waterlink noted [above](#issuecomment-122084697)). That doesn't solve the original problem with [`sandi_meter`](https://github.com/makaroni4/sandi_meter) that led to me filing this issue; if other code can't find my code where it expects to find it, then having aliases stuck off somewhere really doesn't do much good, does it?

In an ideal world, It Would Be Very Nice If you could wrap the original method with a wrapper that maintains the original's name *and parameter count/naming*, but I'm not sure even that would solve my original problem. (Why isn't the wrapper just treated as an intermediate step?)

Feel free to close this; I'm at a loss for how to go forward from here.

**Comment by alex-fedorov** [comment]
@jdickey Thank you for your response. Can you provide a minimal example that shows the problem and how to reproduce? I have some ideas how to fix the problem, but I need something small to play around.

**Comment by waterlink** [comment]
This can be solved if contracts were applied differently:

```ruby
require "contracts/non_intrusive"
C = Contracts

class Example
  include Contracts::NonIntrusive

  Contract(C::Num, C::Num => C::Num)[:add__raw]
  def add__raw(a, b)
    a + b
  end
end
```

This is quite verbose and explicit and you handle mangling yourself. But: it allows static/dynamic analysis tool to find your code correctly and provide useful insights.

Sounds like a topic for another gem (`contracts-non_intrusive` ?).

**Comment by egonSchiele** [comment]
I personally can't imagine using that. Because how the method is named `add__raw` but you call it using `add`...that would be confusing. I'm happy to close this issue as out of scope.

**Comment by waterlink** [comment]
here: https://github.com/waterlink/contracts-non_intrusive - kicking off contracts DSL allowing you to manage method name mangling yourself.
